File: 20_validParentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """

        parentheses = {')':'(','}':'{',']':'['}

        stack = []

        for i in range(len(s)):

            if s[i] not in parentheses:
                stack.append(s[i])
            else:
                if stack != []:
                    if parentheses[s[i]] != stack.pop():
                        return False
                else:
                    return False

        if stack == []:
            return True
        else:
            return False
    

File: 20 Valid Parentheses/20_validParentheses.go
package main

import "fmt"

func isValid(s string) bool {

	// declare empty stack
	var stack []rune

	var parentheses = map[rune]rune{')':'(', ']':'[', '}':'{'}

	for _, value := range s{
		
		// if is closing parenthesis
		if opening, found := parentheses[value]; found{

			if len(stack) == 0 || stack[len(stack)-1] != opening{
				return false
			}

			// remove from stack
			stack = stack[:len(stack)-1]
		} else {

			stack = append(stack,value)

		}

	}

	return len(stack) == 0
    
}

func main() {

	s := "(]"
	fmt.Println(isValid(s))

}